server/neptune/workflows/internal/deploy/workflow.go

```diff
@@ -0,0 +1,132 @@
+package deploy
+
+import (
+	"time"
+
+	"github.com/runatlantis/atlantis/server/neptune/workflows/internal/activities"
+	"github.com/runatlantis/atlantis/server/neptune/workflows/internal/config"
+	"github.com/runatlantis/atlantis/server/neptune/workflows/internal/config/logger"
+	"github.com/runatlantis/atlantis/server/neptune/workflows/internal/deploy/revision/queue"
+	"github.com/runatlantis/atlantis/server/neptune/workflows/internal/deploy/signals"
+	"github.com/runatlantis/atlantis/server/neptune/workflows/internal/github"
+	"go.temporal.io/sdk/workflow"
+)
+
+const (
+	TaskQueue = "deploy"
+
+	RevisionReceiveTimeout = 60*time.Second
+)
+
+type TimedReceiver interface {
+	DidTimeout() bool
+	AddTimeout(ctx workflow.Context, selector workflow.Selector, timeout time.Duration)
+	AddReceiveWithTimeout(ctx workflow.Context, selector workflow.Selector, timeout time.Duration)
+}
+
+type QueueWorker interface {
+	Work(ctx workflow.Context)
+	GetState() queue.WorkerState
+}
+
+func Workflow(ctx workflow.Context, request Request) error {
+	options := workflow.ActivityOptions{
+		TaskQueue:              TaskQueue,
+		ScheduleToCloseTimeout: 5 * time.Second,
+	}
+	ctx = workflow.WithActivityOptions(ctx, options)
+
+	runner := newRunner(ctx, request)
+
+	// blocking call
+	return runner.Run(ctx)
+}
+
+type Runner struct {
+	QueueWorker      QueueWorker
+	Selector         workflow.Selector
+	RevisionReceiver TimedReceiver
+}
+
+func newRunner(ctx workflow.Context, request Request) *Runner {
+	// set relevant logging context
+	ctx = workflow.WithValue(ctx, config.RepositoryLogKey, request.Repository.FullName)
+	ctx = workflow.WithValue(ctx, config.ProjectLogKey, request.Root.Name)
+	ctx = workflow.WithValue(ctx, config.GHRequestIDLogKey, request.GHRequestID)
+
+	// convert to internal types, we should probably move these into another struct
+	repo := github.Repo{
+		Name:     request.Repository.Name,
+		Owner:    request.Repository.Owner,
+		FullName: request.Repository.FullName,
+		URL:      request.Repository.URL,
+	}
+
+	// inject dependencies
+
+	// temporal effectively "injects" this, it just cares about the method names,
+	// so we're modeling our own DI around this.
+	var a *activities.Deploy
+
+	revisionQueue := queue.NewQueue()
+	revisionReceiver := signals.NewRevisionSignalReceiver(ctx, revisionQueue, )
+
+	worker := &queue.Worker{
+		Queue:      revisionQueue,
+		Activities: a,
+		Repo:       repo,
+		RootName:   request.Root.Name,
+	}
+
+	return &Runner{
+		QueueWorker:      worker,
+		RevisionReceiver: revisionReceiver,
+	}
+}
+
+func (r *Runner) Run(ctx workflow.Context) error {
+	workerCtx, cancel := workflow.WithCancel(ctx)
+
+	wg := workflow.NewWaitGroup(ctx)
+	wg.Add(1)
```

Comment:
If we're only kicking off one piece of work, do we even need the waitgroup here?

Reply:
i mean we have this go routine and the main loop, we want to make sure the go routine ends before we shut down the mainloop right?

Reply:
Yeah. I think my thought was more that there are possibly more lightweight versions of doing that, but I acknowledge a waitgroup makes it easy to understand 😄